DllImportGenerator/designs/StructMarshalling.md

```diff
@@ -0,0 +1,127 @@
+
+I propose an opt-in design where the owner of a struct has to explicitly opt-in to usage for interop. This enables our team to add special support as desired for various types such as `Span<T>` while also avoiding the private reflection and limited type information issues mentioned above.
+
+This design would use three attributes:
+
+```csharp
+
+public class GeneratedMarshallingAttribute : Attribute {}
+
+public class BlittableTypeAttribute : Attribute {}
+
+public class NativeStructMarshallingAttribute : Attribute
+{
+     public NativeStructMarshallingAttribute(Type nativeStructType, string managedToNativeMarshalMethod, string nativeToManagedMarshalMethod, string nativeFreeMethod) {}
+}
+
+```
+
+There are 2 usage mechanisms of these attributes.
+
+- Usage 1, Source-generated interop:
+
+The user can apply the `GeneratedMarshallingAttribute` to their structure `S`. The source generator will determine if the type is blittable. If it is blittable, the source generator will generate a partial definition and apply the `BlittableTypeAttribute` to the struct type `S`. Otherwise, it will generate a blittable representation of the struct and apply the `NativeStructMarshallingAttribute` and point it to the blittable representation, a method that marshals the managed value to the native value, a method that marshals a value of the native struct to a managed value, and a method that frees any native resources that were allocated for the native value. The blittable representation can either be generated as a separate top-level type or as a nested type on `S`.
+
+- Usage 2, Manual interop:
+
+The user may want to manually mark their types as marshalable in this system due to specific restrictions in their code base around marshaling specific types that the source generator does not account for. We could also use this internally to support custom types in source instead of in the code generator. In this scenario, the user would apply either the `BlittableTypeAttribute` or the `NativeStructMarshallingAttribute` attribute to their struct type. An analyzer would validate that the struct is blittable if the `BlittableTypeAttribute` is applied or validate that the native struct type is blittable and that the methods provided in the attribute have the correct signatures when the `NativeStructMarshallingAttribute` is applied.
+
+The P/Invoke source generator (as well as the struct source generator when nested struct types are used) would use the `BlittableTypeAttribute` and `NativeStructMarshallingAttribute` to determine how to marshal a value type parameter or field instead of looking at the fields of the struct directly.
```

Comment:
What would happen if the type was marked Blittable but the type contained a non-blittable field?

Reply:
The two main concerns here are:
1) Users can write their own IL
2) Users can disable analyzer warnings

In both scenarios, you could have `struct S { int _field; }` in the reference assembly but `struct S { bool _x; }` in the implementation.
There are also odd scenarios where the user changes the backing field without a dependent assembly recompiling (which would be a breaking change, but which is still something that may need consideration).

Reply:
If the user explicitly marks it as blittable, then the analyzer would produce an error. Otherwise, if the type is marked with `[GeneratedMarshalling]`, it will only have the `BlittableTypeAttribute` attribute when it is blittable.

If someone disables the analyzer and applies `BlittableTypeAttribute` manually (which is very much a "don't do this"/unsupported scenario), then we will not generate any marshalling code and behavior at native boundaries would fall into unspecified behavior. (The P/Invoke source generator would currently fall back to the built-in system based on it's implementation)

Reply:
>The two main concerns here are:
> 1. Users can write their own IL
> 2. Users can disable analyzer warnings

I think these are fair concerns, but I don't think they change much.

Anyone writing their own IL is in a very advanced scenario and this isn't something I think we can solve or address. That being said, I am very much open for making an attempt additionally I could be misunderstanding the concern here.

Disabling analyzer warnings is perfectly acceptable and that is the prerogative of any developer. If the developer knows better then the compiler/analyzer they are okay going it on their own and if the runtime crashes in an interop call that is on them. Consider the case of "use prior to initialize" in C/C++; if the warning was on the error would have been flagged and could have been corrected. Disabling the warning is on the developer.

The rebuttal to that is the case where .NET should be designed to help users avoid those mistakes and in general I agree. However, this is already an advanced scenario so I would prefer to focus on making analyzers avoid false positives for incorrect use - not noisy - and enable experts to produce high quality AOT friendly interop libraries that users can consume so very few need to actually author P/Invokes - see [dotnet/pinvoke](https://github.com/dotnet/pinvoke).